Support health extension for rolling ugrade mode

ArcturusZhang opened the pull request:
Fixes https://github.com/terraform-providers/terraform-provider-azurerm/issues/8120

jackofallops reviewed (changes requested):
Hi @ArcturusZhang 
Thanks for this PR. 
I've left a few comments below based on our conversations offline.

This will need to wait for us to promote the in-line extensions out of beta before being merged as until then the behaviours won't be consistent/predictable with externally defined extensions. We're expecting to review that soon, I'll update here and re-review at that point, but will mark this as blocked for now.

ArcturusZhang commented on azurerm/internal/services/compute/windows_virtual_machine_scale_set_resource.go:
```diff
@@ -419,15 +415,20 @@ func resourceArmWindowsVirtualMachineScaleSetCreate(d *schema.ResourceData, meta
 		}
 	}
 
+	if upgradeMode == compute.UpgradeModeAutomatic && len(automaticOSUpgradePolicyRaw) > 0 && (healthProbeId == "" || !hasHealthExtension) {
```
Sorry that I forgot to add my new checking. I also added some comment about what error message the service will return in the both circumstances.

ArcturusZhang replied:
And sorry I just realized that I made a logic error here - it should be `&&` instead of `||`

ArcturusZhang commented on azurerm/internal/services/compute/virtual_machine_scale_set.go:
```diff
@@ -1367,27 +1367,33 @@ func VirtualMachineScaleSetExtensionsSchema() *schema.Schema {
 }
 
-func expandVirtualMachineScaleSetExtensions(input []interface{}) (*compute.VirtualMachineScaleSetExtensionProfile, error) {
+func expandVirtualMachineScaleSetExtensions(input []interface{}) (*compute.VirtualMachineScaleSetExtensionProfile, bool, error) {
```
Done

ArcturusZhang commented:
Hi @jackofallops I think this might be a part of the extension beta shipping out. When the beta flag is not set, the flag I added will always be `false` and therefore the check
```
(healthProbeId == "" || !hasHealthExtension)
```
will be exactly the same as `healthProbeId == ""` since `!hasHealthExtension` is always `true`, which makes the logic exactly the same as before I changed it. 
What do you think?

marcelkmfst commented:
Are there any information when this PR will be merged / released? It´s a critical feature for us..

marcelkmfst commented:
Hi, is there any update? This is now open for a long time?

Thanks!

sammetb commented:
what is the status about this PR ?
This fix is urgent now

jackofallops commented on azurerm/internal/services/compute/linux_virtual_machine_scale_set_resource.go:
```diff
@@ -433,6 +434,18 @@ func resourceLinuxVirtualMachineScaleSetCreate(d *schema.ResourceData, meta inte
 
+	// otherwise the service return the error:
+	// Automatic OS Upgrade is not supported for this Virtual Machine Scale Set because a health probe or health extension was not specified.
+	if upgradeMode == compute.Automatic && len(automaticOSUpgradePolicyRaw) > 0 && (healthProbeId == "" && !hasHealthExtension) {
```
I don't think we'll get here, line 352 above will kick us out first if `health_probe_id` is unset, so that will need investigating for refactoring.

ArcturusZhang replied:
Yeah... something is messed up during the merging from master. I need to check these carefully again.

jackofallops commented on azurerm/internal/services/compute/windows_virtual_machine_scale_set_resource.go:
```diff
@@ -449,6 +450,18 @@ func resourceWindowsVirtualMachineScaleSetCreate(d *schema.ResourceData, meta in
 
+	// otherwise the service return the error:
+	// Automatic OS Upgrade is not supported for this Virtual Machine Scale Set because a health probe or health extension was not specified.
+	if upgradeMode == compute.Automatic && len(automaticOSUpgradePolicyRaw) > 0 && (healthProbeId == "" && !hasHealthExtension) {
```
As with the Linux resource, we'll not get here if `health_probe_id` is not set, Line 368.

jackofallops reviewed (changes requested):
Hi @ArcturusZhang - Apologies for the long delay in returning to this review. I've left some comments below regarding changes required, if you can address those I'll jump back on it asap.

Thanks!

ArcturusZhang commented on azurerm/internal/services/compute/virtual_machine_scale_set.go:
```diff
@@ -1420,41 +1420,46 @@ func expandVirtualMachineScaleSetExtensions(input []interface{}) (*compute.Virtu
 			protectedSettings, err := structure.ExpandJsonFromString(val.(string))
 			if err != nil {
-				return nil, fmt.Errorf("failed to parse JSON from `protected_settings`: %+v", err)
+				return nil, false, fmt.Errorf("failed to parse JSON from `settings`: %+v", err)
```
My bad, I just corrected this.

ArcturusZhang commented on azurerm/internal/services/compute/linux_virtual_machine_scale_set_extensions_test.go:
```diff
@@ -157,6 +157,36 @@ func TestAccLinuxVirtualMachineScaleSet_extensionsUpdate(t *testing.T) {
 	})
 }
 
+func TestAccAzureRMLinuxVirtualMachineScaleSet_extensionsRollingUpgradeWithHealthExtension(t *testing.T) {
```
Done

ArcturusZhang commented on azurerm/internal/services/compute/linux_virtual_machine_scale_set_extensions_test.go:
```diff
@@ -157,6 +157,36 @@ func TestAccLinuxVirtualMachineScaleSet_extensionsUpdate(t *testing.T) {
+	})
+}
+
+func TestAccAzureRMLinuxVirtualMachineScaleSet_extensionsAutomaticUpgradeWithHealthExtension(t *testing.T) {
```
Done

ArcturusZhang commented on azurerm/internal/services/compute/windows_virtual_machine_scale_set_extensions_test.go:
```diff
@@ -157,6 +157,36 @@ func TestAccWindowsVirtualMachineScaleSet_extensionUpdate(t *testing.T) {
 	})
 }
 
+func TestAccAzureRMWindowsVirtualMachineScaleSet_extensionsRollingUpgradeWithHealthExtension(t *testing.T) {
```
Done

ArcturusZhang commented:
Hi @jackofallops  I have resolved the comments. please take a look again

jackofallops reviewed (approved):
Hi @ArcturusZhang - Thanks for the updates, this LGTM now 👍  I'll get the tests run asap.

jackofallops commented:
Tests looking fine (Failures unrelated)

<img width="666" alt="image" src="https://user-images.githubusercontent.com/11830746/114726386-43b9a700-9d35-11eb-992b-2ea463f6cba1.png">
